session_4/blog/app/controllers/articles_controller.rb

```diff
@@ -0,0 +1,108 @@
+class ArticlesController < ApplicationController
+  load_and_authorize_resource
+
+  before_action :set_article, only: [:show, :edit, :update, :destroy]
+
+  # GET /articles
+  # GET /articles.json
+  def index
+    #to show popular articles
+    @popular_articles=Article.order(views: :DESC).limit(3)
+    #unregistered users can only read 5 articles
+    if session[:user_id] && !session[:private_articles_left]
+      session[:private_articles_left]=current_user.private_articles_left
+    end
+    if !session[:user_id]
+      flash[:notice] ||="Login to read private article"
```

Comment:
Instead of a flash notice, you could do the following - for each private article, disable the show button (or the equivalent) and pop the text "Please login to read private article" to explain why the button is disabled.

https://getbootstrap.com/docs/4.5/components/popovers/